File: leet_code/linked_list/merge_two_sorted_lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/*
  @param {ListNode} l1
  @param {ListNode} l2
  @return {ListNode}
*/

var mergeTwoLists = function (l1, l2) {
    if (!l1 || !l2) return l1 || l2;
    
    let merged = new ListNode();
    // remember to set a pointer to the head of the list
    let head = merged;

    while (l1 && l2) {
        if (l1.val <= l2.val) {
            merged.next = l1;
            l1 = l1.next;
        } else {
            merged.next = l2;
            l2 = l2.next;
        }
        merged = merged.next;
    }

    if (l1) merged.next = l1;
    if (l2) merged.next = l2;

    // make sure to return starting from second node
    return head.next; 
};

File: leet_code/python/linked_list/merge_two_sorted_lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# O(n + m) time, 0(1) space
def mergeTwoLists(self, l1, l2):
    dummy = current = ListNode(0)

    while l1 and l2:
        if l1.val <= l2.val:
            current.next = l1
            l1 = l1.next
            current = current.next
        elif l2.val <= l1.val:
            current.next = l2
            l2 = l2.next
            current = current.next

    current.next = l1 or l2

    return dummy.next
